tests/common/mock_cgroup_commands.py

```diff
@@ -0,0 +1,104 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import contextlib
+import os
+import re
+import subprocess
+
+from azurelinuxagent.common.utils import fileutil
+from tests.tools import patch, data_dir
+
+#
+# Default values for the mocked commands.
+#
+# The output comes from an Ubuntu 18 system
+#
+_default_commands = [
+    (r"systemctl --version",
+'''systemd 237
++PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid
+'''),
+
+    (r"mount -t cgroup",
+'''cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,name=systemd)
+cgroup on /sys/fs/cgroup/rdma type cgroup (rw,nosuid,nodev,noexec,relatime,rdma)
+cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
+cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls,net_prio)
+cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
+cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,nosuid,nodev,noexec,relatime,hugetlb)
+cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
+cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
+cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
+cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
+cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
+cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
+'''),
+
+    (r"mount -t cgroup2",
+'''cgroup on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime)
+'''),
+
+    (r"systemctl show walinuxagent\.service --property CPUAccounting",
+'''CPUAccounting=no
+'''),
+
+    (r"systemctl show walinuxagent\.service --property MemoryAccounting",
+'''MemoryAccounting=no
+'''),
+
+    (r"systemd-run --unit=([^\s]+) --scope ([^\s]+)",
+'''
+Running scope as unit: TEST_UNIT.scope
+Thu 28 May 2020 07:25:55 AM PDT
+''')
+]
+
+_default_files = {
+    "/proc/self/cgroup": os.path.join(data_dir, 'cgroups', 'proc_self_cgroup'),
+    "/sys/fs/cgroup/unified/cgroup.controllers": os.path.join(data_dir, 'cgroups', 'sys_fs_cgroup_unified_cgroup.controllers'),
+}
+
+@contextlib.contextmanager
+def mock_cgroup_commands():
+    original_popen = subprocess.Popen
+    original_read_file = fileutil.read_file
+    original_path_exists = os.path.exists
+
+    def mock_popen(command, *args, **kwargs):
+        if isinstance(command, list):
+            command_string = " ".join(command)
+            for c in _default_commands:
```

Comment:
Curious to know why not just use a dict inplace of a list for `_default_commands`. Would definitely save you the trouble of walking the list everytime

Reply:
initially it was a dict, but then i realized strict string comparison was not enough for I need to do and change it to a sequential list of regular expressions so each time I do a sequential match.

now, most of those regular expressions are simple strings so at some point I considered having a dict search followed by a sequential regex match, but the tests actually run fast enough as it is, so I gave up the idea.